### PR TITLE
[Resource] Instance Refactor

### PR DESCRIPTION
**Pull request by michaelkad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000
Change Summary:
Intended changes were for `resource_ibm_pi_instance` but other resources were affected. Here a quick summary of those other changes, mainly name change to remove duplicate.
- `Attr_Status` replacing `PICloudConnectionStatus, Attr_DhcpStatus, Attr_SharedProcessorPoolStatus`
- `Arg_AffinityPolicy` replacing `PIAffinityPolicy`
- `Arg_AffinityVolume` replacing `PIAffinityVolume`
- `Arg_AffinityInstance` replacing `PIAffinityInstance`
- `Arg_AntiAffinityVolumes` replacing `PIAntiAffinityVolumes`
- `Arg_AntiAffinityInstances` replacing `PIAntiAffinityInstances`
- `Arg_HealthStatus` replacing `Arg_PVMInstanceHealthStatus`
- `StatusOK` replacing `PVMInstanceHealthOk`
- `StatusWarning` replacing `PVMInstanceHealthWarning`
- `Attr_StatusDetail` replacing `Attr_SharedProcessorPoolStatusDetail`
- `Arg_VolumeIDs` replacing `PIVolumeIds`

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'
=== RUN   TestAccIBMPIInstanceBasic
--- PASS: TestAccIBMPIInstanceBasic (516.53s)
PASS


=== RUN   TestAccIBMPIInstanceIBMiLicense
--- PASS: TestAccIBMPIInstanceIBMiLicense (618.72s)
PASS

=== RUN   TestAccIBMPIInstanceReplicant
--- PASS: TestAccIBMPIInstanceReplicant (1239.04s)
PASS

=== RUN   TestAccIBMPIInstanceNetwork
--- PASS: TestAccIBMPIInstanceNetwork (1107.70s)
PASS

=== RUN   TestAccIBMPIInstanceMixedStorage
--- PASS: TestAccIBMPIInstanceMixedStorage (1141.47s)
PASS

=== RUN   TestAccIBMPIInstanceUpdateStoppedState
--- PASS: TestAccIBMPIInstanceUpdateStoppedState (1019.82s)
PASS

=== RUN   TestAccIBMPIInstanceUpdateActiveState
--- PASS: TestAccIBMPIInstanceUpdateActiveState (1088.30s)
PASS

=== RUN   TestAccIBMPIInstanceDeploymentTypeNoStorage

--- PASS: TestAccIBMPIInstanceDeploymentTypeNoStorage (643.53s)
PASS
...
=== RUN   TestAccIBMPISAPInstance
 There is no more capacity to deploy SAP VMs in this data center dal12. Try again later or in a different data center.) Error:bad request Message:
--- FAIL: TestAccIBMPISAPInstance (32.77s)
FAIL

=== RUN   TestAccIBMPIInstanceDeploymentType
Error: failed to create the lpar: Exceeded maximum number of retries. Exhausted all hosts available for retrying build failures for instance
--- FAIL: TestAccIBMPIInstanceDeploymentType (89.97s)
FAIL
```

### CODE REVIEW

File: ibm/service/power/resource_ibm_pi_instance.go [hunk 1]
```diff
@@ -442,10 +443,15 @@ func resourceIBMPIInstanceCreate(ctx context.Context, d *schema.ResourceData, me
 			if err != nil {
 				return diag.FromErr(err)
 			}
+			// TODO: remove this before pushing
+			// _, err = isWaitForPIInstanceSPA(ctx, client, *s.PvmInstanceID, instanceReadyStatus)
+			// if err != nil {
+			// 	return diag.FromErr(err)
+			// }
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_instance.go:450]: Remove this commented out code?

File: ibm/service/power/resource_ibm_pi_instance.go [hunk 2]
```diff
@@ -674,10 +680,10 @@ func resourceIBMPIInstanceUpdate(ctx context.Context, d *schema.ResourceData, me
 
 		//if d.GetOkExists("reboot_for_resource_change")
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_instance.go:681]: Remove commented out code.

File: ibm/service/power/resource_ibm_pi_instance.go [hunk 3]
```diff
@@ -644,25 +650,25 @@ func resourceIBMPIInstanceUpdate(ctx context.Context, d *schema.ResourceData, me
...
 
 	// Start of the change for Memory and Processors
-	if d.HasChange(helpers.PIInstanceMemory) || d.HasChange(helpers.PIInstanceProcessors) {
+	if d.HasChange(Arg_Memory) || d.HasChange(Arg_Processors) {
 
-		maxMemLpar := d.Get("max_memory").(float64)
-		maxCPULpar := d.Get("max_processors").(float64)
+		maxMemLpar := d.Get(Attr_MaxMemory).(float64)
+		maxCPULpar := d.Get(Attr_MaxProcessors).(float64)
 		//log.Printf("the required memory is set to [%d] and current max memory is set to  [%d] ", int(mem), int(maxMemLpar))
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_instance.go:672]: Remove commented out code.

File: ibm/service/power/resource_ibm_pi_instance.go [hunk 4]
```diff
@@ -972,42 +978,42 @@ func isPIInstanceSoftwareLicensesRefreshFunc(client *st.IBMPIInstanceClient, id
 		// Check that each software license we modified has been updated
 		if softwareLicenses.IbmiCSS != nil {
 			if *softwareLicenses.IbmiCSS != *pvm.SoftwareLicenses.IbmiCSS {
-				return pvm, "notdone", nil
+				return pvm, StatusNotDone, nil
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_instance.go:981]: I had put done/notdone, but please feel free to change it if there a status that sounds more appropriate for waiting of the ibmi licenses to update.

File: ibm/service/power/resource_ibm_pi_instance.go [hunk 5]
```diff
@@ -1393,71 +1399,71 @@ func createPVMInstance(d *schema.ResourceData, client *st.IBMPIInstanceClient, i
...
-	if dt, ok := d.GetOk(PIInstanceDeploymentType); ok {
+	if dt, ok := d.GetOk(Arg_DeploymentType); ok {
 		body.DeploymentType = dt.(string)
 	}
 
-	if ap, ok := d.GetOk(PIAffinityPolicy); ok {
+	if ap, ok := d.GetOk(Arg_AffinityPolicy); ok {
 		policy := ap.(string)
 		affinity := &models.StorageAffinity{
 			AffinityPolicy: &policy,
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_instance.go:1433]: Maybe rename this affinity variable to something like "affinityPolicy:. It's already competing names with the constant "Affinity."

**Comment by ismirlia** [comment]
Is the instance resource markdown file going to be added to this PR?

File: ibm/service/power/resource_ibm_pi_instance.go (hunk 5 above)
Comment (michaelkad) [reply ibm/service/power/resource_ibm_pi_instance.go:1428]: I don't think there is any, they are completely different from one another. See below.
`Arg_AffinityPolicy   = "pi_affinity_policy"` vs `Affinity  = "affinity"`

File: website/docs/r/pi_instance.html.markdown [hunk 6]
```diff
@@ -50,95 +54,98 @@ Example usage:
 
 The `ibm_pi_instance` provides the following [timeouts](https://www.terraform.io/docs/language/resources/syntax.html) configuration options:
 
-- **create** - The creation of the instance is considered failed if no response is received for 120 minutes.
-- **Update** The updation of the instance is considered failed if no response is received for 60 minutes.
-- **delete** - The deletion of the instance is considered failed if no response is received for 60 minutes.
-
+* **create** - The creation of the instance is considered failed if no response is received for 120 minutes.
+* **Update** The updation of the instance is considered failed if no response is received for 60 minutes.
```
Comment (ismirlia) [inline website/docs/r/pi_instance.html.markdown:58]:
```suggestion
* **update** The update of the instance is considered to have failed if no response is received for 60 minutes.```

File: website/docs/r/pi_instance.html.markdown [hunk 7]
```diff
@@ -50,95 +54,98 @@ Example usage:
...
+* **delete** - The deletion of the instance is considered failed if no response is received for 60 minutes.
```
Comment (ismirlia) [inline website/docs/r/pi_instance.html.markdown:59]:
```suggestion
* **delete** - The deletion of the instance is considered to have failed if no response is received for 60 minutes.
```

File: website/docs/r/pi_instance.html.markdown [hunk 8]
```diff
@@ -50,95 +54,98 @@ Example usage:
 
 The `ibm_pi_instance` provides the following [timeouts](https://www.terraform.io/docs/language/resources/syntax.html) configuration options:
 
-- **create** - The creation of the instance is considered failed if no response is received for 120 minutes.
-- **Update** The updation of the instance is considered failed if no response is received for 60 minutes.
-- **delete** - The deletion of the instance is considered failed if no response is received for 60 minutes.
-
+* **create** - The creation of the instance is considered failed if no response is received for 120 minutes.
```
Comment (ismirlia) [inline website/docs/r/pi_instance.html.markdown:57]:
```suggestion
* **create** - The creation of the instance is considered to have failed if no response is received for 120 minutes.
```

**Review by ismirlia — APPROVED** [review]

**Comment by michaelkad** [comment]
Opened on main repo

**Pull request closed without merging**